src/PowerShellEditorServices/Services/DebugAdapter/DebugService.cs

```diff
@@ -456,22 +457,25 @@ public async Task<string> SetVariableAsync(int variableContainerReferenceId, str
 
             if (argTypeConverterAttr is not null)
             {
+                // PSVariable *is* strongly typed, so we have to convert it.
                 _logger.LogTrace($"Setting variable '{name}' using conversion to value: {expressionResult ?? "<null>"}");
 
-                // TODO: This is throwing a 'PSInvalidOperationException' thus causing
-                // 'DebuggerSetsVariablesWithConversion' to fail.
-                psVariable.Value = await _executionService.ExecuteDelegateAsync(
-                    "PS debugger argument converter",
-                    ExecutionOptions.Default,
-                    (pwsh, _) =>
-                    {
-                        var engineIntrinsics = (EngineIntrinsics)pwsh.Runspace.SessionStateProxy.GetVariable("ExecutionContext");
-
-                        // TODO: This is almost (but not quite) the same as LanguagePrimitives.Convert(), which does not require the pipeline thread.
-                        //       We should investigate changing it.
-                        return argTypeConverterAttr.Transform(engineIntrinsics, expressionResult);
-                    },
+                // NOTE: We use 'Get-Variable' here instead of 'SessionStateProxy.GetVariable()'
+                // because we already have a pipeline running (the debugger) and the latter cannot
+                // run concurrently (threw 'NoSessionStateProxyWhenPipelineInProgress').
+                IReadOnlyList<EngineIntrinsics> results = await _executionService.ExecutePSCommandAsync<EngineIntrinsics>(
+                    new PSCommand()
+                        .AddCommand(@"Microsoft.PowerShell.Utility\Get-Variable")
+                        .AddParameter("Name", "ExecutionContext")
+                        .AddParameter("ValueOnly"),
                     CancellationToken.None).ConfigureAwait(false);
+                EngineIntrinsics engineIntrinsics = results.Count > 0
+                    ? results[0]
+                    : throw new Exception("Couldn't get EngineIntrinsics!");
+
+                // TODO: This is almost (but not quite) the same as 'LanguagePrimitives.Convert()',
+                // which does not require the pipeline thread. We should investigate changing it.
```

Comment:
I looked into this, but I don't think it will quite as "well" as the transformation, and I'm unsure actually how to go about it because [`LanguagePrimitives.ConvertTo()`](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.languageprimitives.convertto?view=powershellsdk-7.0.0#System_Management_Automation_LanguagePrimitives_ConvertTo_System_Object_System_Type_System_IFormatProvider_) doesn't accept an `ArgumentTypeConverterAttribute`.

Reply:
This will only work for local debugging sessions.  Does this code make local/remote distinctions?

Reply:
It does not make that distinction. Also, do you mean that `LanguagePrimitives` would only work for local, or the currently proposed change?

Reply:
The ExecutionContext variable (engineIntrinsics) passes through the serialization system for the remoting case, and is just a PSObject property bag and not a System.Management.Automation.EngineIntrinsics type.  The serialization system does not rehydrate returned object into their original types, except for a small set of special cases.

So if you have script code that determines variable scope, that script should run entirely on the remote target, so that it can run in the correct context and 'live' type instances.

Reply:
I'm not quite sure I'm following. Can I ask: will the proposed change behave _differently_ than the existing code in the remoting scenario? If not, then let's kick that down the road a bit.

Reply:
(I.e. are both implementations broken in the same way?)

Reply:
Yes, I think so (remoting may be broken).  But I need to look more closely at the code (and not just the changes).